### PR TITLE
Premodit

### PR DESCRIPTION
**Pull request by HajimeKawahara**
First implementation done.
I reorganize the test directories 

- `tests/unittests` was made for the unit tests. 
- `tests/integration` for the integration test
- `tests/endtoend` for e2e tests/

I suppose that the reviewers at least run the unit tests. 
Also, in this PR, please check the reverse modeling by

``` sh
python tests/endtoend/reverse_premodit.py
```

Here is my output. It took ~ 1 hour using a laptop PC.

``` sh
/home/kawahara/anaconda3/lib/python3.8/site-packages/exojax-1.1.3-py3.8.egg/exojax/spec/rtransfer.py:14: UserWarning: rtransfer.nugrid is duplicated and will be removed. Use `setrt.gen_wavenumber_grid` instead
  warnings.warn(warn_msg, UserWarning)
xsmode assumes ESLOG: mode= premodit
WARNING: resolution may be too small. R= 561053.1928844501
Background atmosphere:  H2
Reading transition file
.broad is used.
Broadening code level= a1
default broadening parameters are used for  23  J lower states in  40  states
H2-H2
N= 53638606
uniqidx: 100%|██████████████████████████████████████████████████████████████████████████████████████████| 2/2 [00:00<00:00,  2.04it/s]
sample: 100%|██████████████████████████████████████████| 3000/3000 [53:28<00:00,  1.07s/it, 33 steps of size 6.76e-03. acc. prob=0.88]

                mean       std    median      5.0%     95.0%     n_eff     r_hat
   MMR_CH4      0.01      0.00      0.01      0.00      0.01      7.89      1.01
        RV     10.66      0.29     10.67     10.15     11.12     19.12      1.00
        Rp      0.89      0.07      0.88      0.72      0.98     12.63      1.00
        T0   1313.33     10.98   1312.36   1298.38   1334.28     24.94      1.00
     alpha      0.10      0.00      0.10      0.10      0.10     19.16      1.15
     vsini     20.11      0.56     20.13     19.24     21.10      8.47      1.15

Number of divergences: 1952

```

Figure_1x
Figure_1y

### CODE REVIEW

File: src/exojax/spec/modit.py [hunk 1]
```diff
@@ -314,18 +219,30 @@ def hitran(mdb, Tarr, Parr, Pself, R, molmass):
...
     return SijM, ngammaLM, nsigmaDl
 
 
-def setdgm_hitran(mdb, fT, Parr, Pself_ref, R, molmass, res, *kargs):
+def setdgm_hitran(mdb, fT, Parr, Pself_ref, R, molmass, dit_grid_resolution,
+                  *kargs):
+    warn_msg = "`modit.setdgm_hitran` is duplicated and will be removed. Use `modit.set_ditgrid_matrix_hitran` instead"
+    warnings.warn(warn_msg, UserWarning)
+    return set_ditgrid_matrix_hitran(mdb, fT, Parr, Pself_ref, R, molmass,
+                                     dit_grid_resolution, kargs)
```
Comment (chonma0ctopus) [inline src/exojax/spec/modit.py:241]: kargs -> *kargs

File: src/exojax/spec/modit.py [hunk 2]
```diff
@@ -430,12 +350,25 @@ def vald_all(asdb, Tarr, PH, PHe, PHH, R):
...
 def setdgm_vald_each(ielem, iion, atomicmass, ionE, dev_nu_lines, logsij0, elower, eupper, gamRad, gamSta, vdWdamp, \
-                QTmask, T_gQT, gQT_284species, PH, PHe, PHH, R, fT, res, *kargs):  # (SijM, ngammaLM, nsigmaDl, fT, res, *kargs):
+                QTmask, T_gQT, gQT_284species, PH, PHe, PHH, R, fT, dit_grid_resolution, *kargs):
+    warn_msg = "`modit.setdgm_vald_each` is duplicated and will be removed. Use `modit.set_ditgrid_matrix_vald_each` instead"
+    warnings.warn(warn_msg, UserWarning)
+    return set_ditgrid_matrix_vald_each(ielem, iion, atomicmass, ionE,
+                                        dev_nu_lines, logsij0, elower, eupper,
+                                        gamRad, gamSta, vdWdamp, QTmask, T_gQT,
+                                        gQT_284species, PH, PHe, PHH, R, fT,
+                                        dit_grid_resolution, kargs)
```
Comment (chonma0ctopus) [inline src/exojax/spec/modit.py:365]: Same as above (Thank you for updating the metal-related functions as well!)

File: src/exojax/spec/modit.py [hunk 3]
```diff
@@ -467,18 +400,29 @@ def setdgm_vald_each(ielem, iion, atomicmass, ionE, dev_nu_lines, logsij0, elowe
...
+        set_dgm_minmax, dit_grid_resolution=dit_grid_resolution)
     return jnp.array(dgm_ngammaL)
 
 
-def setdgm_vald_all(asdb, PH, PHe, PHH, R, fT, res, *kargs):
+def setdgm_vald_all(asdb, PH, PHe, PHH, R, fT, dit_grid_resolution, *kargs):
+    warn_msg = "`modit.setdgm_vald_all` is duplicated and will be removed. Use `modit.set_ditgrid_matrix_vald_all` instead"
+    warnings.warn(warn_msg, UserWarning)
+    return set_ditgrid_matrix_vald_all(asdb, PH, PHe, PHH, R, fT,
+                                       dit_grid_resolution, kargs)
```
Comment (chonma0ctopus) [inline src/exojax/spec/modit.py:421]: Same as above

File: src/exojax/spec/moldb.py [hunk 4]
```diff
@@ -109,8 +110,7 @@ def __init__(self, path, nurange=[-np.inf, np.inf], margin=0.0, crit=0., Ttyp=10
         pf = exomolapi.read_pf(self.pf_file)
         self.gQT = jnp.array(pf['QT'].to_numpy())  # grid QT
         self.T_gQT = jnp.array(pf['T'].to_numpy())  # T forgrid QT
-        self.Tref = 296.0
```
Comment (chonma0ctopus) [inline src/exojax/spec/moldb.py:112]: self.Tref -> Tref in [line 765 (in func. qr_iso_interp)](https://github.com/HajimeKawahara/exojax/blob/c476903f8c2034d0b793a461b7dcc033be173a74/src/exojax/spec/moldb.py#L765)

File: src/exojax/spec/modit.py [hunk 5]
```diff
@@ -467,18 +400,29 @@ def setdgm_vald_each(ielem, iion, atomicmass, ionE, dev_nu_lines, logsij0, elowe
...
              QTmask, ielem, iion, atomicmass, ionE, \
                    dev_nu_lines, logsij0, elower, eupper, gamRad, gamSta, vdWdamp)
-        floop = lambda c, arr:  (c, jnp.nan_to_num(arr, nan=jnp.nanmin(arr)))
+        floop = lambda c, arr: (c, jnp.nan_to_num(arr, nan=jnp.nanmin(arr)))
         ngammaLM = scan(floop, 0, ngammaLM)[1]
-        set_dgm_minmax.append(minmax_dgmatrix(ngammaLM, res))
-    dgm_ngammaL = precompute_dgmatrix(set_dgm_minmax, res=res)
+        set_dgm_minmax.append(
+            minmax_ditgrid_matrix(ngammaLM, dit_grid_resolution))
+    dgm_ngammaL = precompute_ditgrid_matrix(
```
Comment (chonma0ctopus) [inline src/exojax/spec/modit.py:412]: precompute_ditgrid_matrix -> precompute_modit_ditgrid_matrix
[Line 276](https://github.com/HajimeKawahara/exojax/blob/16b08dc01067e52402a6a17007a955e277dff601/src/exojax/spec/modit.py#L276) (in set_ditgrid_matrix_hitran) may be the same.

File: tests/manual_check/premodit/compare_with_direct3D.py [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+def compare_with_direct3D(mdb,nus,Ttest=1000.0,interval_contrast=0.1,Ttyp=2000.0):
+    """ compare the premodit LSD with the direct computation of LSD 3D version
+    
+    """
+    from exojax.spec.lsd import npgetix
+    from exojax.spec.hitran import SijT
+    from exojax.spec.premodit import make_elower_grid, unbiased_lsd
+
+    broadpar=np.array([mdb._n_Texp,mdb._alpha_ref]).T
+    uidx_broadpar, uniq_broadpar=uniqidx_2D(broadpar)
```
Comment (chonma0ctopus) [inline tests/manual_check/premodit/compare_with_direct3D.py:16]: uniqidx_2D is not defined (exojax.utils.indexing.uniqidx maybe?). 
Some other functions below such as npadd1D or make_lbd3D_uniqidx are also missing.
(Do I not need to check this test...?)

Comment (HajimeKawahara) [reply tests/manual_check/premodit/compare_with_direct3D.py]: This test is not used now. Thanks!

File: src/exojax/spec/premodit.py [hunk 7]
```diff
@@ -0,0 +1,338 @@
...
+from exojax.spec import normalized_doppler_sigma
+
+
+@jit
+def xsvector(T, P, nsigmaD, lbd, R, pmarray, nu_grid, elower_grid,
+             multi_index_uniqgrid, ngamma_ref_grid, n_Texp_grid, qt):
+    """compute cross section vector
+
+    Args:
+        T (_type_): temperature in Kelvon
```
Comment (chonma0ctopus) [inline src/exojax/spec/premodit.py:22]: Kelvin

**Review by chonma0ctopus — APPROVED** [review]
badge for the implementation! I added very minor comments. The other parts look great to me.

File: src/exojax/spec/initspec.py [hunk 8]
```diff
@@ -38,16 +43,16 @@ def init_dit(nu_lines, nu_grid, warning=False):
         pmarray: (+1,-1) array whose length of len(nu_grid)+1
 
     Note:
-       cont is the contribution for i=index. 1 - cont is the contribution for i=index+1. For other i, the contribution should be zero.
+        cont is the contribution for i=index. 1 - cont is the contribution for i=index+1. For other i, the contribution should be zero.
```
[the remark this answers is not in the document]
Comment (HajimeKawahara) [reply src/exojax/spec/initspec.py:46]: Exactly! Thanks.

**Comment by HajimeKawahara** [comment]
Thanks! Corrected typos.

**Comment by ykawashima** [comment]
Thank you very much for the implementation, and sorry for the delay in reviewing.
I have confirmed that tests/endtoend/reverse_premodit.py works!
However, I have realized that the following three tests do not pass in my environment.
- tests/unittests/spec/modit/modit_test.py
- tests/unittests/spec/premodit/premodit_spectrum_test.py
- tests/unittests/install/func.py (this is just due to my server environment?)

Do you have any solutions?

tests/unittests/spec/modit/modit_test.py
```
(base) [kawashima@higuma modit]$ python modit_test.py 
/home/kawashima/anaconda3/lib/python3.9/site-packages/pkg_resources/__init__.py:123: PkgResourcesDeprecationWarning: 4.0.0-unsupported is an invalid version and will not be supported in a future release
  warnings.warn(
xsmode assumes ESLOG: mode= modit
/home/kawashima/anaconda3/lib/python3.9/site-packages/exojax-1.1.3-py3.9.egg/exojax/spec/initspec.py:200: UserWarning: Some of the line centers are outside of the wavenumber grid.
  warnings.warn(
/home/kawashima/anaconda3/lib/python3.9/site-packages/exojax-1.1.3-py3.9.egg/exojax/spec/initspec.py:202: UserWarning: All of the line center should be within wavenumber grid for PreMODIT/MODIT/DIT.
  warnings.warn(
0.029209018
Traceback (most recent call last):
  File "/home/kawashima/ExoJAX/check/unittests/spec/modit/modit_test.py", line 107, in <module>
    test_rt_exomol()
  File "/home/kawashima/ExoJAX/check/unittests/spec/modit/modit_test.py", line 100, in test_rt_exomol
    assert np.all(residual < 0.02)
AssertionError
```

tests/unittests/spec/premodit/premodit_spectrum_test.py
```
(base) [kawashima@higuma premodit]$ python premodit_spectrum_test.py
/home/kawashima/anaconda3/lib/python3.9/site-packages/pkg_resources/__init__.py:123: PkgResourcesDeprecationWarning: 4.0.0-unsupported is an invalid version and will not be supported in a future release
  warnings.warn(
xsmode assumes ESLOG: mode= modit
/home/kawashima/anaconda3/lib/python3.9/site-packages/exojax-1.1.3-py3.9.egg/exojax/spec/initspec.py:200: UserWarning: Some of the line centers are outside of the wavenumber grid.
  warnings.warn(
/home/kawashima/anaconda3/lib/python3.9/site-packages/exojax-1.1.3-py3.9.egg/exojax/spec/initspec.py:202: UserWarning: All of the line center should be within wavenumber grid for PreMODIT/MODIT/DIT.
  warnings.warn(
uniqidx: 100%|█████████████████████████████████████████| 1/1 [00:00<00:00, 15709.00it/s]
0.02941066
Traceback (most recent call last):
  File "/home/kawashima/ExoJAX/check/unittests/spec/premodit/premodit_spectrum_test.py", line 75, in <module>
    test_rt_exomol()
  File "/home/kawashima/ExoJAX/check/unittests/spec/premodit/premodit_spectrum_test.py", line 70, in test_rt_exomol
    assert np.all(residual < 0.02)
AssertionError
```

tests/unittests/install/func.py
```(base) [kawashima@higuma install]$ python func.py 
2022-07-17 19:28:39.281129: W external/org_tensorflow/tensorflow/stream_executor/platform/default/dso_loader.cc:64] Could not load dynamic library 'libcudnn.so.8'; dlerror: libcudnn.so.8: cannot open shared object file: No such file or directory
2022-07-17 19:28:39.281171: E external/org_tensorflow/tensorflow/stream_executor/cuda/cuda_dnn.cc:372] Could not create cudnn handle: CUDNN_STATUS_INTERNAL_ERROR
2022-07-17 19:28:40.003772: E external/org_tensorflow/tensorflow/stream_executor/cuda/cuda_dnn.cc:372] Could not create cudnn handle: CUDNN_STATUS_INTERNAL_ERROR
2022-07-17 19:28:40.004170: E external/org_tensorflow/tensorflow/stream_executor/cuda/cuda_dnn.cc:372] Could not create cudnn handle: CUDNN_STATUS_INTERNAL_ERROR
Traceback (most recent call last):
  File "/home/kawashima/ExoJAX/check/unittests/install/func.py", line 21, in <module>
    test_cuDNN()
  File "/home/kawashima/ExoJAX/check/unittests/install/func.py", line 10, in test_cuDNN
    c = jnp.convolve(s, v, mode='same')
  File "/home/kawashima/anaconda3/lib/python3.9/site-packages/jax/_src/traceback_util.py", line 162, in reraise_with_filtered_traceback
    return fun(*args, **kwargs)
  File "/home/kawashima/anaconda3/lib/python3.9/site-packages/jax/_src/api.py", line 416, in cache_miss
    out_flat = xla.xla_call(
  File "/home/kawashima/anaconda3/lib/python3.9/site-packages/jax/core.py", line 1632, in bind
    return call_bind(self, fun, *args, **params)
  File "/home/kawashima/anaconda3/lib/python3.9/site-packages/jax/core.py", line 1623, in call_bind
    outs = primitive.process(top_trace, fun, tracers, params)
  File "/home/kawashima/anaconda3/lib/python3.9/site-packages/jax/core.py", line 1635, in process
    return trace.process_call(self, fun, tracers, params)
  File "/home/kawashima/anaconda3/lib/python3.9/site-packages/jax/core.py", line 627, in process_call
    return primitive.impl(f, *tracers, **params)
  File "/home/kawashima/anaconda3/lib/python3.9/site-packages/jax/interpreters/xla.py", line 687, in _xla_call_impl
    compiled_fun = _xla_callable(fun, device, backend, name, donated_invars,
  File "/home/kawashima/anaconda3/lib/python3.9/site-packages/jax/linear_util.py", line 263, in memoized_fun
    ans = call(fun, *args)
  File "/home/kawashima/anaconda3/lib/python3.9/site-packages/jax/interpreters/xla.py", line 759, in _xla_callable_uncached
    return lower_xla_callable(fun, device, backend, name, donated_invars,
  File "/home/kawashima/anaconda3/lib/python3.9/site-packages/jax/interpreters/xla.py", line 892, in compile
    self._executable = XlaCompiledComputation.from_xla_computation(
  File "/home/kawashima/anaconda3/lib/python3.9/site-packages/jax/interpreters/xla.py", line 921, in from_xla_computation
    compiled = compile_or_get_cached(backend, xla_computation, options)
  File "/home/kawashima/anaconda3/lib/python3.9/site-packages/jax/interpreters/xla.py", line 863, in compile_or_get_cached
    return backend_compile(backend, computation, compile_options)
  File "/home/kawashima/anaconda3/lib/python3.9/site-packages/jax/interpreters/xla.py", line 474, in backend_compile
    return backend.compile(built_c, compile_options=options)
jax._src.traceback_util.UnfilteredStackTrace: RuntimeError: UNKNOWN: Failed to determine best cudnn convolution algorithm: UNKNOWN: GetConvolveAlgorithms failed.

Convolution performance may be suboptimal.  To ignore this failure and try to use a fallback algorithm, use XLA_FLAGS=--xla_gpu_strict_conv_algorithm_picker=false.  Please also file a bug for the root cause of failing autotuning.

The stack trace below excludes JAX-internal frames.
The preceding is the original exception that occurred, unmodified.

--------------------

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/home/kawashima/ExoJAX/check/unittests/install/func.py", line 21, in <module>
    test_cuDNN()
  File "/home/kawashima/ExoJAX/check/unittests/install/func.py", line 10, in test_cuDNN
    c = jnp.convolve(s, v, mode='same')
  File "/home/kawashima/anaconda3/lib/python3.9/site-packages/jax/interpreters/xla.py", line 474, in backend_compile
    return backend.compile(built_c, compile_options=options)
RuntimeError: UNKNOWN: Failed to determine best cudnn convolution algorithm: UNKNOWN: GetConvolveAlgorithms failed.

Convolution performance may be suboptimal.  To ignore this failure and try to use a fallback algorithm, use XLA_FLAGS=--xla_gpu_strict_conv_algorithm_picker=false.  Please also file a bug for the root cause of failing autotuning.```

**Comment by HajimeKawahara** [comment]
@ykawashima 

Thank you! So, both the residuals of modit and premodit exceed 2% (it's almost ~3%). I can fix these test by changing to 
```python
assert np.all(residual < 0.03)
```
but before that can you check the difference by making figures?
```python
plt.plot(nus,F0[:12000])
plt.plot(dat["nus"],dat["flux"].values[:12000])
```

**Comment by ykawashima** [comment]
Here are the plots. Indeed, I can recognize the differences.
tests/unittests/spec/modit/modit_test.py
```
    plt.plot(nus[:12000],F0[:12000])
    plt.plot(dat["nus"].values[:12000],dat["flux"].values[:12000])
    plt.show()
```
modit

tests/unittests/spec/premodit/premodit_spectrum_test.py
```
    plt.plot(nu_grid[:12000],F0[:12000])
    plt.plot(dat["nus"].values[:12000],dat["flux"].values[:12000])
    plt.show()
```
premodit

File: tests/unittests/spec/modit/modit_test.py [hunk 9]
```diff
@@ -0,0 +1,107 @@
...
+    xsm = xsmatrix(cont_nu, index_nu, R, pmarray, nsigmaDl, ngammaLM, SijM,
+                   nus, dgm_ngammaL)
+    dtau = dtauM(dParr, jnp.abs(xsm), MMR * np.ones_like(Parr), molmass, g)
+    sourcef = piBarr(Tarr, nus)
+    F0 = rtrun(dtau, sourcef)
+    filename = pkg_resources.resource_filename(
+        'exojax', 'data/testdata/' + TESTDATA_CO_EXOMOL_MODIT_EMISSION_REF)
+    dat = pd.read_csv(filename, delimiter=",", names=("nus", "flux"))
+    
+    # The reference data was generated by
```
Comment (HajimeKawahara) [inline tests/unittests/spec/modit/modit_test.py:91]: check. here

**Comment by HajimeKawahara** [comment]
Thanks! We postpone the last issue in #266.

**Pull request merged**